include/iterlib/Item.h

```diff
@@ -0,0 +1,205 @@
+
+  ItemOptimized(const ItemOptimized& other) = default;
+  ItemOptimized& operator=(const ItemOptimized& other) = default;
+  ItemOptimized& operator=(ItemOptimized&& other) = default;
+
+  void reset() {
+    id_ = 0;
+    ts_ = 0;
+    *this = dynamic();
+  }
+
+  id_t id() const { return id_; }
+
+  void setId(id_t id) { id_ = id; }
+
+  void setTs(int64_t ts) { ts_ = ts; }
+
+  int64_t ts() const { return ts_; }
+
+  void syncIdTs() {
+    try {
+      id_ = at(QueryConstants::kIdKey).get<int64_t>();
+      ts_ = at(QueryConstants::kTimeKey).get<int64_t>();
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  // This depends on value().which(). For eg: vector_pair_t can't
+  // be modified.
+  void syncMap() {
+    try {
+      (*this)[QueryConstants::kIdKey] = dynamic(int64_t(id_));
+      (*this)[QueryConstants::kTimeKey] = dynamic(int64_t(ts_));
+    } catch (const std::exception& ex) {
+      LOG_EVERY_N(WARNING, 1000) << ex.what();
+    }
+  }
+
+  void sync() {
+    // This is a heuristic. The actual sync depends on what was
+    // modified. Use syncIdTs/syncMap directly depending on the
+    // context.
+    if (id_ == 0) {
+      syncIdTs();
+    } else {
+      syncMap();
+    }
+  }
+
+  const dynamic& value() const { return *this; }
+
+  bool operator<(const ItemOptimized& other) const {
+    return std::tie(value(), ts_, id_) <
+           std::tie(other.value(), other.ts_, other.id_);
+  }
+
+  bool operator==(const ItemOptimized& other) const {
+    return std::tie(value(), id_, ts_) ==
+           std::tie(other.value(), other.id_, other.ts_);
+  }
+
+  // Weaker form of operator== which ignores types.
+  // May relax the equality criteria further based on use cases.
+  bool equals(const ItemOptimized& other) const {
+    if (std::tie(id_, ts_) != std::tie(other.id_, other.ts_)) {
+      return false;
+    }
+    if (value().empty() && other.value().empty()) {
+      return true;
+    }
+    return value() == other.value();
+  }
+
+  // Used by groupby. Note that id/ts are not populated
+  ItemOptimized& operator=(const dynamic& other) {
+    dynamic::operator=(other);
+    return *this;
+  }
+
+  ItemOptimized& operator=(dynamic&& other) {
+    dynamic::operator=(other);
+    return *this;
+  }
+
+  bool isAdditionalField(const std::string& key) const {
+    return key == kIdKey || key == kTimeKey;
+  }
+
+  int64_t getAdditionalField(const std::string& key) const {
+    if (key == kIdKey) {
+      return id_;
+    } else if (key == kTimeKey) {
+      return ts_;
+    }
+    throw std::logic_error("not an additional field");
+  }
+
+  std::string toString() const {
+    return folly::sformat("[{0}, {1}, {2}]", ts(), id(), dynamic::toString());
+  }
+
+ protected:
+  id_t id_;
+  int64_t ts_;
+};
+
+// partialCompare is a 3-value function:
+//
+// LT  => less than
+// EQ  => equal
+// GT_OR_NOT_COMPARABLE  => uncomparable or greater than
```

Comment:
Because I wasn't brave enough to make the changes in one step. Will break it into two.